File: tests/CoreTests/CoreTests/HTMLCodegen/SimpleTests.fs
module Tests.SimpleTests

open Expecto
open Plotly.NET
open Plotly.NET.LayoutObjects
open Plotly.NET.TraceObjects


open TestUtils.HtmlCodegen
open Chart2DTestCharts

[<Tests>]
let ``Html layout tests`` =
    testList "SimpleTests.Simple tests" [
        testCase "Expecting plotly js script reference in generated html document" ( fun () ->
            $"""https://cdn.plot.ly/plotly-{Globals.PLOTLYJS_VERSION}.min"""
            |> substringIsInChart Chart2DTestCharts.Point.``Point chart with axis labels and title`` GenericChart.toEmbeddedHTML 
        );
        testCase "Expecting data" ( fun () ->
            """var data = [{"type":"scatter","mode":"markers","x":[0.0,1.0,2.0,3.0,4.0,5.0,6.0,7.0,8.0,9.0,10.0],"y":[0.0,1.0,2.0,3.0,4.0,5.0,6.0,7.0,8.0,9.0,10.0],"marker":{},"line":{}}];"""
            |> chartGeneratedContains Chart2DTestCharts.Point.``Point chart with axis labels and title``
        );
        testCase "Expecting layout info" (fun () ->
            """var layout = {"title":{"text":"Hello world!"},"xaxis":{"title":{"text":"xAxis"},"showgrid":false},"yaxis":{"title":{"text":"yAxis"},"showgrid":false}};"""
            |> chartGeneratedContains Chart2DTestCharts.Point.``Point chart with axis labels and title``
        );
        testCase "Expecting html tags in embedded page only" (fun () ->
            ["<html>"; "</html>"; "<head>"; "</head>"; "<body>"; "</body>"; "<script type=\"text/javascript\">"; "</script>"]
            |> substringListIsInChart Chart2DTestCharts.Point.``Point chart with axis labels and title`` GenericChart.toEmbeddedHTML
        );
        testCase "Expecting some html tags in both embedded and not embedded" (fun () ->
            ["<script type=\"text/javascript\">"; "</script>"]
            |> chartGeneratedContainsList Chart2DTestCharts.Point.``Point chart with axis labels and title``
        );
        testCase "Passing args to the function" ( fun () ->
            "data, layout, config);"
            |> chartGeneratedContains Chart2DTestCharts.Point.``Point chart with axis labels and title``
        )
    ]


[<Tests>]
let ``plotlyjs reference tests`` =
    testList "SimpleTests.plotlyjs reference" [
        testCase "full reference" (fun _ ->
            getFullPlotlyJS()
            |> substringIsInChart Chart2DTestCharts.Point.``Point chart with full plotly.js reference`` GenericChart.toEmbeddedHTML
        )
        testCase "cdn reference" (fun _ ->
            $"""https://cdn.plot.ly/plotly-2.0.0.min"""
            |> substringIsInChart Chart2DTestCharts.Point.``Point chart with plotly.js cdn reference`` GenericChart.toEmbeddedHTML 
        )
        testCase "require: Expecting cloudflare link" (fun () ->
            """https://cdnjs.cloudflare.com/ajax/libs/require.js"""
            |> chartGeneratedContains Chart2DTestCharts.Point.``Point chart referencing plotly.js using require.js``
        );
        testCase "require: Expecting require config" (fun () ->
            (sprintf """var fsharpPlotlyRequire = requirejs.config({context:'fsharp-plotly',paths:{plotly:'https://cdn.plot.ly/plotly-%s.min'}}) || require;""" Globals.PLOTLYJS_VERSION)
            |> chartGeneratedContains Chart2DTestCharts.Point.``Point chart referencing plotly.js using require.js``
        );
    ]

File: tests/ExtensionLibsTests/CSharpTests/htmlcodegen/Chart2D/SimpleTests.cs
using System;
using System.Linq;
using Xunit;
using Plotly.NET.CSharp;

// avoid the namespace 'Plotly.NET', as we dont want to accidentally use F# type extensions instead of real C# bindings
namespace CSharp.Tests.HtmlCodegen.Chart2D
{
    public class SimpleTests
    {
        private const string ExpectedData = """var data = [{"type":"scatter","mode":"markers","x":[0.0,1.0,2.0,3.0,4.0,5.0,6.0,7.0,8.0,9.0,10.0],"y":[0.0,1.0,2.0,3.0,4.0,5.0,6.0,7.0,8.0,9.0,10.0],"marker":{},"line":{}}];""";

        private const string ExpectedLayout = """var layout = {"xaxis":{"title":{"text":"xAxis"}},"yaxis":{"title":{"text":"yAxis"}}};""";

        static Plotly.NET.GenericChart simpleChart =
            Chart.Point<double, double, string>(
                x: Enumerable.Range(0, 11).Select(x => Convert.ToDouble(x)).ToArray(),
                y: Enumerable.Range(0, 11).Select(x => Convert.ToDouble(x)).ToArray(),
                UseDefaults: false
            )
                .WithXAxisStyle<int, int, int>(TitleText: "xAxis")
                .WithYAxisStyle<int, int, int>(TitleText: "yAxis");

        public class HtmlLayoutTests
        {
            [Fact]
            public void PlotlyJsScriptInGeneratedHTMLDocument()
            {
                TestUtils.SubstringIsInChart(
                    simpleChart,
                    Plotly.NET.GenericChart.toEmbeddedHTML, 
                    $"https://cdn.plot.ly/plotly-{Globals.PLOTLYJS_VERSION}.min"
                );
            }

            [Fact]
            public void ExpectingData()
            {
                TestUtils.ChartGeneratedSectionEquals(
                    simpleChart,
                    ChartMarkupSection.Data,
                    ExpectedData
                );
            }

            [Fact]
            public void ExpectingLayoutInfo()
            {
                TestUtils.ChartGeneratedSectionEquals(
                    simpleChart,
                    ChartMarkupSection.Layout,
                    ExpectedLayout
                );
            }

            [Fact]
            public void ExpectingHTMLTagsInEmbeddedPageOnly()
            {
                string[] tags = new string[] { "<html>", "</html>", "<head>", "</head>", "<body>", "</body>", "<script type=\"text/javascript\">", "</script>" };
                TestUtils.SubstringListIsInChart(
                    simpleChart,
                    Plotly.NET.GenericChart.toEmbeddedHTML,
                    tags
                );
            }

            [Fact]
            public void ExpectingSomeHTMLTagsInBothEmbeddedAndNotEmbedded()
            {
                string[] tags = new string[] { "<script type=\"text/javascript\">", "</script>" };
                TestUtils.ChartGeneratedContainsList(
                    simpleChart,
                    tags
                );
            }

            [Fact]
            public void PassingArgsToTheFunction()
            {
                TestUtils.ChartGeneratedContains(
                    simpleChart,
                    "data, layout, config);"
                );
            }

        }
    }
}
